File: plugin-sdk-ts/src/batch.ts
/**
 * Batch operations module.
 *
 * Provides batch install, uninstall, and update operations
 * as well as package history and pinning.
 *
 * Reuses existing `pkg_search` and `pkg_install` permissions.
 */
import { callHostJson } from './host';
import type {
  BatchItem,
  BatchResult,
  PackageHistoryEntry,
  PinnedPackage,
  UpdateInfo,
} from './types';

/** Batch install packages. Requires: pkg_install */
export function batchInstall(items: BatchItem[]): BatchResult {
  return callHostJson<BatchResult>(
    'cognia_batch_install',
    JSON.stringify({ items }),
  );
}

/** Batch uninstall packages. Requires: pkg_install */
export function batchUninstall(items: BatchItem[]): BatchResult {
  return callHostJson<BatchResult>(
    'cognia_batch_uninstall',
    JSON.stringify({ items }),
  );
}

/** Batch update packages. Requires: pkg_install */
export function batchUpdate(items: BatchItem[]): BatchResult {
  return callHostJson<BatchResult>(
    'cognia_batch_update',
    JSON.stringify({ items }),
  );
}

/** Check for updates across packages. Requires: pkg_search */
export function checkUpdates(packages: string[], provider: string): UpdateInfo[] {
  return callHostJson<UpdateInfo[]>(
    'cognia_batch_check_updates',
    JSON.stringify({ packages, provider }),
  );
}

/** Get package operation history. Requires: pkg_search */
export function getHistory(limit?: number): PackageHistoryEntry[] {
  return callHostJson<PackageHistoryEntry[]>(
    'cognia_batch_get_history',
    JSON.stringify({ limit: limit ?? null }),
  );
}

/** Get list of pinned packages. Requires: pkg_search */
export function getPinnedPackages(): PinnedPackage[] {
  return callHostJson<PinnedPackage[]>('cognia_batch_get_pinned', '');
}

File: plugin-sdk/src/batch.rs
//! Batch operations module.
//!
//! Provides batch install, uninstall, update, history, and pinning.

use crate::host;
use crate::types::*;
use extism_pdk::*;

/// Batch install packages. Requires: pkg_install
pub fn batch_install(items: &[BatchItem]) -> Result<BatchResult, Error> {
    let input = serde_json::json!({ "items": items }).to_string();
    let result = unsafe { host::cognia_batch_install(input)? };
    Ok(serde_json::from_str(&result)?)
}

/// Batch uninstall packages. Requires: pkg_install
pub fn batch_uninstall(items: &[BatchItem]) -> Result<BatchResult, Error> {
    let input = serde_json::json!({ "items": items }).to_string();
    let result = unsafe { host::cognia_batch_uninstall(input)? };
    Ok(serde_json::from_str(&result)?)
}

/// Batch update packages. Requires: pkg_install
pub fn batch_update(items: &[BatchItem]) -> Result<BatchResult, Error> {
    let input = serde_json::json!({ "items": items }).to_string();
    let result = unsafe { host::cognia_batch_update(input)? };
    Ok(serde_json::from_str(&result)?)
}

/// Check for updates across packages. Requires: pkg_search
pub fn check_updates(packages: &[&str], provider: &str) -> Result<Vec<UpdateInfo>, Error> {
    let input = serde_json::json!({ "packages": packages, "provider": provider }).to_string();
    let result = unsafe { host::cognia_batch_check_updates(input)? };
    Ok(serde_json::from_str(&result)?)
}

/// Get package operation history. Requires: pkg_search
pub fn get_history(limit: Option<u32>) -> Result<Vec<PackageHistoryEntry>, Error> {
    let input = serde_json::json!({ "limit": limit }).to_string();
    let result = unsafe { host::cognia_batch_get_history(input)? };
    Ok(serde_json::from_str(&result)?)
}

/// Get list of pinned packages. Requires: pkg_search
pub fn get_pinned_packages() -> Result<Vec<PinnedPackage>, Error> {
    let result = unsafe { host::cognia_batch_get_pinned(String::new())? };
    Ok(serde_json::from_str(&result)?)
}
